### PR TITLE
Fix serialization for type which has DataMember with special Name values

### PR DESCRIPTION
**Pull request by khdang**
This issue happens only for DCJS and affecting both NetCore and NetNative, due to missing method from recent merging of DCJS.

cc: @SGuyGe @shmao @zhenlan

### CODE REVIEW

**Comment by shmao** [comment]
LGTM.

**Comment by stephentoub** [comment]
Should we add Debug.Asserts in all of the places that access such methods via reflection, e.g.
https://github.com/dotnet/corefx/blob/dfe8f72f7f4cbf61f08cf71b4e9f7c455c1e3dbf/src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonFormatGeneratorStatics.cs#L404-L405
to make sure that this doesn't happen again (or isn't already happening elsewhere)?

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs [hunk 1]
```diff
@@ -326,6 +326,12 @@ private void VerifyType(DataContract dataContract, Type declaredType)
                 scopedKnownTypes.Pop();
             }
         }
+
+        internal static void WriteJsonNameWithMapping(XmlWriterDelegator xmlWriter, XmlDictionaryString[] memberNames, int index)
```
Comment (weshaggard) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: Who calls this method?

Comment (weshaggard) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: Nevermind @stephentoub already pointed me to who calls it. I do agree with trying to add some validation where we can.

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: It's accessed via reflection (I had to search for it, as I had a similar question in mind); see my comment for a link to the access site.

Comment (khdang) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: I've added Debug.Assert to all applicable places that access to type members through reflection. I think there is validation check in NetNative toolchain but it shows as warning and not failure.

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlObjectSerializerWriteContextComplexJson.cs]: Thanks for adding those, looks good.

**Pull request merged**